plugins/rebalance/rest.go

```diff
@@ -0,0 +1,323 @@
+
+	"github.com/gorilla/mux"
+	"github.com/pborman/uuid"
+	log "github.com/sirupsen/logrus"
+)
+
+func createRebalanceInfo(volname string, req *rebalanceapi.StartReq) *rebalanceapi.RebalInfo {
+	return &rebalanceapi.RebalInfo{
+		Volname:     volname,
+		RebalanceID: uuid.NewRandom(),
+                Status:      rebalanceapi.Started,
+		Cmd:         checkCmd(req),
+		CommitHash:  setCommitHash(),
+		RebalStats:       rebalanceapi.RebalNodeStatus{},
+	}
+}
+
+func rebalanceStartHandler(w http.ResponseWriter, r *http.Request) {
+
+	ctx := r.Context()
+	logger := gdctx.GetReqLogger(ctx)
+
+	// collect inputs from url
+	volname := mux.Vars(r)["volname"]
+
+	var req rebalanceapi.StartReq
+
+//  TODO : How do I pass these options?
+//  Unmarshal Request so to handle fix-layout and start force
+
+//      if err := restutils.UnmarshalRequest(r, &req); err != nil {
+//              restutils.SendHTTPError(ctx, w, http.StatusBadRequest, errors.ErrJSONParsingFailed)
+//              return
+//	}
+
+	vol, err := volume.GetVolume(volname)
+	if err != nil {
+		restutils.SendHTTPError(ctx, w, http.StatusNotFound, errors.ErrVolNotFound)
+		return
+	}
+
+	if vol.State != volume.VolStarted {
+		restutils.SendHTTPError(ctx, w, http.StatusBadRequest, errors.ErrVolNotStarted)
+		return
+	}
+
+	if vol.DistCount == 1 {
+		restutils.SendHTTPError(ctx, w, http.StatusBadRequest, errors.ErrVolNotDistribute)
+		return
+	}
+
+	// Check for remove- brick pending
+	//TODO
+
+	// A simple transaction to start rebalance on all nodes
+	txn := transaction.NewTxn(ctx)
+	defer txn.Cleanup()
+
+	lock, unlock, err := transaction.CreateLockSteps(volname)
+	if err != nil {
+		restutils.SendHTTPError(r.Context(), w, http.StatusInternalServerError, err)
+		return
+	}
+
+	txn.Nodes = vol.Nodes()
+	txn.Steps = []*transaction.Step{
+		lock,
+		{
+			DoFunc: "rebalance-start",
+			Nodes:  txn.Nodes,
+		},
+		{
+			DoFunc: "rebalance-store",
+			Nodes:  []uuid.UUID{gdctx.MyUUID},
+		},
+		unlock,
+	}
+
+	err = txn.Ctx.Set("volname", volname)
+	if err != nil {
+		logger.WithError(err).Error("failed to set volname in transaction context")
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+		return
+	}
+
+	rebalinfo := createRebalanceInfo(volname, &req)
+	if rebalinfo == nil {
+		logger.WithError(err).Error("failed to create Rebalance info")
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+		return
+	}
+
+	err = txn.Ctx.Set("rinfo", rebalinfo)
+	if err != nil {
+		logger.WithError(err).Error("failed to set rebalance info in transaction context")
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+		return
+	}
+
+	err = txn.Do()
+	if err != nil {
+		/* TODO: Need to handle failure case. Unlike other process,
+		 * rebalance process is one per node and depends on volfile change.
+		 * Need to handle scenarios where process is started in
+		 * few nodes and failed in few others */
```

Comment:
If you create and set undo-funcs in the transaction steps, they should automatically take care of stopping any started rebalance processes.